### PR TITLE
Add support for extra arguments such as authentication object

### PR DESCRIPTION
**Pull request by akelge**
This patch adds support for authentication to the client.
It uses the same logic as `requests` (http://docs.python-requests.org/en/master/user/advanced/#custom-authentication).

More I added a small fix to keep in account cases when `links` can be empty (i.e. Flask-Restless pagination)

### CODE REVIEW

File: src/jsonapi_client/session.py [hunk 1]
```diff
@@ -119,10 +119,13 @@ class Session:
     def __init__(self, server_url: str=None,
                  enable_async: bool=False,
                  schema: dict=None,
+                 auth: object=None,
                  loop: 'AbstractEventLoop'=None) -> None:
         self._server: ParseResult
         self.enable_async = enable_async
 
+        self.authObject = auth
```
Comment (tuomas2) [inline src/jsonapi_client/session.py]: Please rename this variable to auth_object (PEP8)

**Comment by tuomas2** [comment]
Thank you @akelge for this pull request! 

- Could you please document this feature in README.rst
- Did you test both synchronous (requests) as well as asyncio API?

**Comment by akelge** [comment]
Here we are!
I renamed the variable to adhere to PEP8.
About async I don't know how to test it, this is the first time I use python 3 :)
Your examples do not work on python 3.6.

But I am quite confident it should work as, basically, I used the requests system to add a callback to the authenticator on each request.

**Comment by akelge** [comment]
I forgot to add documentation, done now!

**Comment by tuomas2** [comment]
I made a little change to approach. Instead of supporting authentication object, i support now any extra arguments that you might like to pass to request.request(). 

Reason for not liking supporting only authentication object was that I found out that API for requests and aiohttp are different so the documented example would have been working only for requests. And now it is more general, allowing passing special cookies etc. to the server if needed.

**Comment by tuomas2** [comment]
> Your examples do not work on python 3.6.

What do you mean by that? jsonapi-client works only with python 3.6+

**Comment by akelge** [comment]
Yes, you are right, your approach is better, I like it.
About tests, as I told you this is the first time I adventure in python 3 lands. I am quite experienced with python 2, but I am learning the new stuff of python 3 (and fighting with strings management, eheheheh).

I tried copy/pasting your example about async usage into ipython and I got some nasty "Syntax error" messages. I have found that the async/await couple has been introduced with python 3.5, so I supposed something changed between 3.5 and 3.6.
Now I am on the urge to close a project quickly, later on I will investigate/study and find why and how to make it work in my test envo.

**Comment by akelge** [comment]
I fear that your last commit introduced a regression.
Checking your code it seems that self.href was set to empty string, no matter what was inside data.
That was breaking json parsing in case of empty lists.
Now it should work better

**Comment by tuomas2** [comment]
> I tried copy/pasting your example about async usage into ipython and I got some nasty "Syntax error" messages. I have found that the async/await couple has been introduced with python 3.5, so I supposed something changed between 3.5 and 3.6.
Now I am on the urge to close a project quickly, later on I will investigate/study and find why and how to make it work in my test envo.

You need to run them within an asyncio event loop. You can learn more about asyncio in https://docs.python.org/3/library/asyncio.html . It wont work directly from ipython shell. 

> I fear that your last commit introduced a regression.
Checking your code it seems that self.href was set to empty string, no matter what was inside data.
That was breaking json parsing in case of empty lists.
Now it should work better

Oops, sorry I indeed introduced and error there. Thanks for fixing it.

**Comment by tuomas2** [comment]
I think this is good to go now.

**Pull request merged**